lib/Parse/ParseExpr.cpp

```diff
@@ -3023,6 +3022,74 @@ ExprResult Parser::ParseBoundsExpression() {
   return Result;
 }
 
+// Recurse through types and substitute any TypedefType with TypeVariableType
+// as the underlying type.
+QualType Parser::SubstituteTypeVariable(QualType QT,
+  SmallVector<DeclRefExpr::GenericFunctionCallInfo::TypeNameInfo, 16> &typeNames) {
+  const Type *T = QT.getTypePtr();
+  switch (T->getTypeClass()) {
+  case Type::Pointer: {
+    const PointerType *pType = dyn_cast<PointerType>(T);
```

Comment:
We try to practice "defensive programming" in our changes in clang, more so than usually seen. After a dyn_cast, we usually check that the result is non-null as expected.   If it is null, we use llvm_unreachable and return some sort of sensible type.   The other uses of dyn_cast in this function should be updated too.